Tasks/ContainerStructureTestV0/testrunner.ts

```diff
@@ -15,12 +16,23 @@ export class TestRunner {
             try {
                 const runnerDownloadUrl = this.getContainerStructureTestRunnerDownloadPath(this.osType);
                 if (!runnerDownloadUrl) {
-                    throw new Error("Unable to get runner download path");
+                    throw new Error(`Not supported OS: ${this.osType}`);
                 }
 
-                const runnerPath = await this.downloadTestRunner(runnerDownloadUrl);
-                tl.debug(`Successfully downloaded : ${runnerDownloadUrl}`);
+                let toolPath = toolLib.findLocalTool(this.toolName, "1.0.0");
 
+                if(!toolPath) {
+                    const downloadPath = await toolLib.downloadTool(runnerDownloadUrl);
+                    tl.debug(`Successfully downloaded : ${downloadPath}`);
+                    toolPath = await toolLib.cacheFile(downloadPath, this.toolName, this.toolName, "1.0.0");
+                    tl.debug(`Successfully Added to cache`);
+                } else {
+                    tl.debug(`Tool is retrieved from cache.`);
+                }
+
+                const runnerPath = path.join(toolPath, this.toolName);
+                existsSync(runnerPath);
```

Comment:
can you follow https://github.com/microsoft/azure-pipelines-tasks/blob/a8fd5bc0e8aefec098d20ad37f264a64bc9ed21a/Tasks/Common/kubernetes-common-v2/kubectlutility.ts#L45 this pattern. It may possible cache path may not contain tool so we need to copy if it not exist.

Reply:
Why do we copy again ? does that mean the cacheFile can throw error or malfunction and we are doing fallback?

Reply:
if exe not present in toolpath then we are doing copy.